internal/graphicsdriver/metal/graphics_darwin.go

```diff
@@ -230,13 +231,9 @@ struct FragmentShaderImpl {
     if (useColorM) {
       c.rgb /= c.a + (1.0 - sign(c.a));
       c = (color_matrix_body * c) + color_matrix_translation;
-      c *= v.color;
       c.rgb *= c.a;
```

Comment:
In theory it is still possible that RGB values exceed the alpha value after the calculation. So I think calling `min` is still required for the color-matrix case. Or this will break the compatibility.

Reply:
On second thought, even without a color matrix, this might change the result when RGB values are extremely bigger than alpha. `min` might be required for such case.

Hmm, let me think more. I now started to think that we can break such a subtle backward compatibility...

Reply:
The min should only be needed for roundoff errors, malformed color inputs, or for malformed color matrices.

In theory none of these should ever happen, as linear combinations prefer wellformed premultiplied alpha.

If this cannot be done, though, I would still want to request this as an optional feature for games to opt into (i.e. "I promise my color matrices are all good, now skip this check".

Reply:
Another option of course is to only apply the min in the color matrix case, which has a lot of other complexity the min() can "hide behind" anyway.

Reply:
Here's the primary case where this would change things:

Use a Scale(2.0, 2.0, 2.0, 1.0) color matrix.

This can, on a color like "50% white" (0.5, 0.5, 0.5, 0.5), get you a "50% brighter-than-white" color (1.0, 1.0, 1.0, 0.5).

Now this would still kinda render "correctly" - it'd behave exactly like mixing with an overbright color - but it'd indeed look _different_ and could change games. This can be happening in real world code.

In DarkPlaces engine, we actually did this intentionally - used a premultiplied-alpha blending mode (GL_ONE, GL_ONE_MINUS_SRC_ALPHA), and in the texture we happily mixed normal premultiplied alpha data, and data where color values are set but alpha is all zero to get additive blending.

So, I suppose for a minor release, we probably should prefer keeping the min() there at least in the color matrix case. Do we still need it in the color multiply case though? In that case, I'll refactor this a little so my change only affects the non-color matrix case, and add a TODO to later reconsider or document the behavior of out-of-range colors with a color matrix.

Reply:
Hm... wait. The user doesn't even specify whether colormatrix is to be used or not.

So if I bring back the clamping only to the colormatrix case, one could now easily write code that looks different with and without this clamping (e.g. can be reproduced by using a mostly-scale color matrix vs a scale-only color matrix).

So personally I'd vote for removing this clamping entirely, but if this cannot be done safely, can we get an option to do so?

Reply:
Actually, I found an option: if any color multipliers are > 1, do not invoke the ScaleOnly optimization. I'll update this patch to do that.